core/src/commonMain/kotlin/com/powersync/attachments/AttachmentQueue.kt

```diff
@@ -0,0 +1,489 @@
+            localStorage,
+            attachmentsService,
+            ::getLocalUri,
+            errorHandler,
+            logger,
+            syncThrottleDuration,
+        )
+
+    private var syncStatusJob: Job? = null
+    private val mutex = Mutex()
+
+    public var closed: Boolean = false
+
+    /**
+     * Initialize the attachment queue by
+     * 1. Creating attachments directory
+     * 2. Adding watches for uploads, downloads, and deletes
+     * 3. Adding trigger to run uploads, downloads, and deletes when device is online after being offline
+     */
+    @Throws(PowerSyncException::class, CancellationException::class)
+    public suspend fun startSync(): Unit =
+        runWrappedSuspending {
+            mutex.withLock {
+                if (closed) {
+                    throw Exception("Attachment queue has been closed")
+                }
+                // Ensure the directory where attachments are downloaded, exists
+                localStorage.makeDir(attachmentsDirectory)
+
+                subdirectories?.forEach { subdirectory ->
+                    localStorage.makeDir(Path(attachmentsDirectory, subdirectory).toString())
+                }
+
+                val scope = CoroutineScope(Dispatchers.IO)
+
+                syncingService.startSync(syncInterval)
+
+                // Listen for connectivity changes
+                syncStatusJob =
+                    scope.launch {
+                        val statusJob =
+                            launch {
+                                var previousConnected = db.currentStatus.connected
+                                db.currentStatus.asFlow().collect { status ->
+                                    if (!previousConnected && status.connected) {
+                                        syncingService.triggerSync()
+                                    }
+                                    previousConnected = status.connected
+                                }
+                            }
+
+                        val watchJob =
+                            launch {
+                                // Watch local attachment relationships and sync the attachment records
+                                watchedAttachments.collect { items ->
+                                    processWatchedAttachments(items)
+                                }
+                            }
+
+                        statusJob.join()
+                        watchJob.join()
+                    }
+            }
+        }
+
+    /**
+     * Stops syncing. Syncing may be resumed with [startSync].
+     */
+    @Throws(PowerSyncException::class, CancellationException::class)
+    public suspend fun stopSyncing(): Unit =
+        runWrappedSuspending {
+            mutex.withLock {
+                if (closed) {
+                    return@runWrappedSuspending
+                }
+
+                syncStatusJob?.cancel()
+                syncStatusJob?.join()
```

Comment:
Might need an additional import

```suggestion
                syncStatusJob?.cancelAndJoin()
```

(same thing in `close()`

Reply:
Noice. Implemented this.